### ISSUE TITLE
SNMPv3 not working with APC UPS

### ISSUE THREAD

**Issue by mschaf** [swisscom/ruby-netsnmp#67]
Hello, 
I am trying to pull data of a APC Smart UPS X with the AP9631 network card. 

I can use snmpget to query oids just fine, e.g. 
```
$ snmpget -v3 -u apc -a MD5 -A 123456789 -x AES -X 123456789 -l authPriv 10.2.0.60 -On 1.3.6.1.2.1.33.1.4.4.1.2.1
> .1.3.6.1.2.1.33.1.4.4.1.2.1 = INTEGER: 231 RMS Volts
```
But when I try it via ruby
```
NETSNMP::Client.new(host: "10.2.0.60", username: "apc", auth_password: "123456789", auth_protocol: :md5, priv_password: "12
3456789", priv_protocol: :aes, security_level: :auth_priv).get("1.3.6.1.2.1.33.1.4.4.1.2.1")
```
I get `Timeout after 2 seconds (Timeout::Error`.
My assumtion is, that these two should do exactly the same, am I wrong with that?

I looked at the requests with wireshark and they are slightly different.  
snmpget:  
![Screenshot from 2022-05-05 01-08-09](https://user-images.githubusercontent.com/32179648/166839767-5ba64f25-601f-40aa-938c-ca17db3a4db1.png)
ruby-netsnmp:
![Screenshot from 2022-05-05 01-08-39](https://user-images.githubusercontent.com/32179648/166839786-c935d1fe-512c-48ee-98ba-cde183aa5a5d.png)

I am using:
* Ruby: 2.7.5
* ruby-netsnmp: 0.6.3
* snmpget: NET-SNMP version: 5.8
* Ubuntu 20.04.3 LTS

**Comment by HoneyryderChuck** [comment]
Thx for the report. Can you run your example again in debug mode, as per the instructions in the realmente, and post the output here?

**Comment by HoneyryderChuck** [comment]
The parameter difference does not seem to be the problem, at least from what I'm looking (although `netsnmp` could  omit username from the probe request). The probe response not arriving on time seems to be the problem. Did you try increasing the `:timeout`, let's say, to 10, and see whether it's sent back?

**Comment by mschaf** [comment]
Increasing the timeout did not help. With snmpget the ups responds very fast.
I found that the ups is logging unauthorized snmp access attempts, but nothing there is nothing when using netsnmp.
It just returns nothing, I looked at the network at the switch via port mirroring.
However I can access other equipment on the same network with netsnmp just fine.

**Comment by HoneyryderChuck** [comment]
in v3, the client needs to send a "probe request" (a report PDU) and get a response as part of the initial encryption negotiation, before sending the actual GET request.

with `netsnmp`, the probe is being sent, but no probe response comes back (hence the timeout).

Your wireshark log is not complete, because it does include this initial probe request . Or at least this is my impression by looking at your wireshark log. Can you show the logs for all relevant PDUs? (report request, report response, get request, get response, don't forget to obfuscate actual data I should not be looking at).

**Comment by mschaf** [comment]
This is everthing I captured, just that single packet, I assume this is the probe request.
I will do some digging into the RFCs on the weekend. I guess my best bet is comparing it to the snmpget one and try and get them to be equal until it works.

**Comment by HoneyryderChuck** [comment]
Equality will be possible, due to the randomness involved in encrypting. Maybe what you can do is try with v2, if you can enable it in the box for testing, and see if it works. If it doesn't, there's Smth to do at the network layer. Just hard to figure out next steps without testing it myself.

**Comment by mschaf** [comment]
The ups only has v1 and v3, v1 works just fine.
I will give me the weekend to read into the protocol and debug this. 
After that I think I will go down another path, I have not found a good alternative, so far. I am currently leaning towards parsing snmpget :see_no_evil:

**Comment by HoneyryderChuck** [comment]
snmpget supports  [verbose debug output](https://net-snmp.sourceforge.io/wiki/index.php/Debugging_output), if that helps you with the troubleshooting further

**Comment by mschaf** [comment]
I figured it out, the ups does not like the `msgAuthenticationParameters` of the probe request to be 12 zeros but wants it empty. I think this complies with the RFCs, but I am not 100% sure, those are quite hard to read. snmpget does also not include the 12 zeros. I created a pull request with a fix, if you want to include this change.

**Comment by HoneyryderChuck** [comment]
@mschaf just released v0.6.4 with the patch. Sorry for the slow turnaround time, been a busy couple of days.
